src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -608,103 +610,111 @@ private async Task ProcessRequests<TContext>(IHttpApplication<TContext> applicat
 
                 InitializeBodyControl(messageBody);
 
-                var context = application.CreateContext(this);
+                // We run user controlled request processing in a seperate async method
+                // so any changes made to ExecutionContext are undone when it returns and
+                // each request starts with a fresh ExecutionContext state.
+                await ProcessRequest(application);
 
-                try
-                {
-                    KestrelEventSource.Log.RequestStart(this);
-
-                    // Run the application code for this request
-                    await application.ProcessRequestAsync(context);
-
-                    // Trigger OnStarting if it hasn't been called yet and the app hasn't
-                    // already failed. If an OnStarting callback throws we can go through
-                    // our normal error handling in ProduceEnd.
-                    // https://github.com/aspnet/KestrelHttpServer/issues/43
-                    if (!HasResponseStarted && _applicationException == null && _onStarting?.Count > 0)
-                    {
-                        await FireOnStarting();
-                    }
-
-                    if (!_connectionAborted && !VerifyResponseContentLength(out var lengthException))
-                    {
-                        ReportApplicationError(lengthException);
-                    }
-                }
-                catch (BadHttpRequestException ex)
+                // Even for non-keep-alive requests, try to consume the entire body to avoid RSTs.
+                if (!_connectionAborted && _requestRejectedException == null && !messageBody.IsEmpty)
                 {
-                    // Capture BadHttpRequestException for further processing
-                    // This has to be caught here so StatusCode is set properly before disposing the HttpContext
-                    // (DisposeContext logs StatusCode).
-                    SetBadRequestState(ex);
-                    ReportApplicationError(ex);
+                    await messageBody.ConsumeAsync();
                 }
-                catch (Exception ex)
+
+                if (HasStartedConsumingRequestBody)
                 {
-                    ReportApplicationError(ex);
+                    await messageBody.StopAsync();
                 }
+            }
+        }
 
-                KestrelEventSource.Log.RequestStop(this);
+        private async ValueTask ProcessRequest<TContext>(IHttpApplication<TContext> application)
```

Comment:
`Task` is a more performant return here than `ValueTask` and will allocate the same; i.e none in the sync completion and the async box in the async wait scenario (subject to https://github.com/dotnet/coreclr/pull/26310 when `ValueTask` becomes better as it doesn't allocate in async)